File: July-LeetCoding-Challenge-2020/Week2/662-maximum-width-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 * Runtime: 76 ms, faster than 76.96% of JavaScript online submissions for Maximum Width of Binary Tree.
 * Memory Usage: 39.2 MB, less than 6.12% of JavaScript online submissions for Maximum Width of Binary Tree.
 */
const mins = [0];
let max = 0;

var widthOfBinaryTree = function(root) {
    dfs(root, 0, 0);

    return max;
};

function dfs(v, lvl, pos) {
    if (!v) {
        return;
    }

    if (lvl === mins.length) {
        mins[lvl] = pos;
    }

    const delta = pos - mins[lvl];
    max = Math.max(max, delta + 1);

    dfs(v.left, lvl + 1, delta * 2);
    dfs(v.right, lvl + 1, delta * 2 + 1);
}

function widthOfBinaryTree(root) {
    const mins = [0];
    let max = 0;

    dfs(root, 0, 0);
    return max;

    function dfs(v, lvl, pos) {
        if (!v) return;
        if (lvl === mins.length) {
            mins[lvl] = pos;
        }

        const delta = pos-mins[lvl];
        max = Math.max(max, delta+1);
        dfs(v.left, lvl+1, delta*2);
        dfs(v.right, lvl+1, delta*2+1);
    }
}

File: July-LeetCoding-Challenge-2020/Week2/662-maximum-width-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def widthOfBinaryTree(self, root: TreeNode) -> int:
        if not root:
            return 0
        q = deque([(root,0)])
        width = 0
        while q:
            print("l", left)
            print("r",right)
            width = max(width,right-left+1)
            next_level = deque()
            while q:
                node, index = q.popleft() # popleft so it does traverse from left side (preorder - root, left, right)
                if node.left: 
                    next_level.append((node.left, index*2)) # left subtree root has always 2n index
                if node.right:
                    next_level.append((node.right, index*2+1)) # right subtree root has 2n+1 index number
                    
            q = next_level
        return width
        # TC: O(n + sigma(n-1)) -> O(n^2)
        # SC: O(n), if we exclude queue O(1)